src/Voron/Impl/Compaction/StorageCompaction.cs

```diff
@@ -230,57 +230,97 @@ private static void CopyTrees(StorageEnvironment existingEnv, StorageEnvironment
             // Load table into structure 
             var inputTable = txr.OpenTable(schema, treeName);
 
-            // Create the new table, and replay inserts
-            using (var txw = compactedEnv.WriteTransaction())
-            {
-                schema.Create(txw, treeName);
-                var outputTable = txw.OpenTable(schema, treeName);
+            // The next three variables are used to know what our current
+            // progress is
+            var entriesLeft = inputTable.NumberOfEntries;
+
+            // It is very important that these slices be allocated in the
+            // txr.Allocator, as the intermediate write transactions on
+            // the compacted environment will be destroyed between each
+            // loop.
+            var lastSlice = Slices.BeforeAllKeys;
+            long lastFixedIndex = 0L;
 
-                if (schema.Key == null)
+            while (entriesLeft > 0)
+            {
+                using (var txw = compactedEnv.WriteTransaction())
                 {
-                    // There is no primary key, however, there must be at least one index
-                    if (schema.Indexes.Count > 0)
+                    long transactionSize = 0L;
+
+                    schema.Create(txw, treeName);
+                    var outputTable = txw.OpenTable(schema, treeName);
+
+                    if (schema.Key == null)
                     {
-                        // We have a variable size index, use it
-                        var index = schema.Indexes.First().Value;
+                        // There is no primary key, however, there must be at least one index
+                        if (schema.Indexes.Count > 0)
+                        {
+                            // We have a variable size index, use it
+                            var index = schema.Indexes.First().Value;
 
-                        foreach (var result in inputTable.SeekForwardFrom(index, Slices.BeforeAllKeys))
+                            foreach (var result in inputTable.SeekForwardFrom(index, lastSlice))
+                            {
+                                foreach (var entry in result.Results)
+                                {
+                                    // The table will take care of reconstructing indexes automatically
+                                    outputTable.Insert(entry);
+                                    entriesLeft--;
+                                    transactionSize += entry.Size;
+                                }
+
+                                lastSlice = result.Key;
```

Comment:
Here you are keeping track of the lastSlice after the transaction has been completed, which is bad.